### DISCUSSION TITLE
Justify text written with TexWriter's fill_textbox

### DISCUSSION THREAD

**Discussion by dantard** [pymupdf/PyMuPDF#3522] [Looking for help] [answered]
Hello, I'm trying to understand how `TEXT_ALIGN_JUSTIFY` works. Consider this code:

```
  doc = pymupdf.open()
  page = doc.new_page(-1,  # insertion point: end of document
                      width=595,  # page dimension: A4 portrait
                      height=842)
  font = pymupdf.Font("helv")
  tw = pymupdf.TextWriter(page.rect)
  r = pymupdf.Rect(100, 100, 340, 120)
  page.add_rect_annot(r)
  tw.fill_textbox(r,
                  "Hello, world! Shouldn't this text be justified?",
                  align=pymupdf.TEXT_ALIGN_JUSTIFY, font=font, fontsize=11)
  tw.write_text(page)
  doc.save("doc-with-new-blank-page.pdf")
```

Shouldn't the text be justified to fill the rectangle?
(Notice that the annotation is there just to show the size of the rectangle)

Otherwise, is there any other way to justify the text or to set the font stretch value?

Thanks in advance.

**Comment by JorjMcKie** [comment]
The last line of text will never be justified to prevent awkward appearances.

I suggest you use https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_htmlbox and employ HTML / CSS syntax to achieve special text effects.

**Comment by JorjMcKie** [comment, marked as answer]
For examples using your own font with insert_htmlbox see here: https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-specify-your-own-fonts
